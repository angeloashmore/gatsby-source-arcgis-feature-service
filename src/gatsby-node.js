import url from 'url'
import got from 'got'
import polylabel from 'polylabel'

// Default parameters for the ArcGIS feature server request: fetch all features
// and fields in GeoJSON format.
const DEFAULT_PARAMS = {
  f: 'geojson',
  where: '1=1',
  outFields: '*',
}

// ArcGIS feature type name.
const FEATURE_TYPE = 'ArcGisFeature'

export const sourceNodes = async (gatsbyContext, pluginOptions) => {
  const { actions, createNodeId, createContentDigest } = gatsbyContext
  const { createTypes, createNode } = actions
  const { name, url: serverURL, params } = pluginOptions

  const resolvedURL = url.resolve(serverURL, '0/query')
  const response = await got(resolvedURL, {
    json: true,
    query: { ...DEFAULT_PARAMS, ...params },
  })

  // Set the geometry field as JSON. This field may have a different shape per
  // feature and will likely not need direct access to child properties via
  // GraphQL.
  createTypes(`
    type ${FEATURE_TYPE} implements Node {
      geometry: JSON!
    }
  `)

  // Create ArcGisFeature nodes for each feature.
  response?.body?.features?.forEach?.(feature =>
    createNode({
      ...feature,
      id: createNodeId([name, feature?.id].filter(Boolean).join(' ')),
      featureId: feature?.id,
      polylabel:
        feature.geometry.type === 'Polygon'
          ? polylabel(feature.geometry.coordinates, 0.1)
          : null,
      sourceName: name,
      internal: {
        type: FEATURE_TYPE,
        contentDigest: createContentDigest(feature),
      },
    }),
  )
}
