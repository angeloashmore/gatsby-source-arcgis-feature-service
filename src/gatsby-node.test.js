import got from 'got'
import * as plugin from './gatsby-node'

const mockFeature = {
  id: 'id',
  geometry: {
    type: 'MultiPolygon',
    coordinates: [
      [
        [
          [0, 0],
          [10, 0],
          [10, 10],
          [0, 10],
        ],
      ],
      [
        [
          [0, 0],
          [10, 0],
          [10, 10],
          [0, 10],
        ],
      ],
    ],
  },
  // Arbitrary properties from the ArcGIS feature service.
  properties: {
    foo: 'bar',
    bar: 'baz',
  },
  // Ensure stray properties are **not** included.
  foo: 'bar',
}

// got is used to fetch JSON data from the ArcGIS Feature service. We need to
// mock its return value in tests.
jest.mock('got')
got.mockImplementation((url, options) => {
  // Ensure we are telling got to parse JSON response data.
  if (!options.json) throw new Error('Did not request JSON data from server.')

  // Ensure we are requesting GeoJSON data from the server.
  if (options.query.f !== 'geojson')
    throw new Error('Did not request GeoJSON data from server.')

  return { body: { features: [mockFeature] } }
})

const gatsbyContext = {
  actions: {
    createNode: jest.fn(),
    createTypes: jest.fn(),
  },
  createNodeId: jest.fn().mockReturnValue('createNodeId'),
  createContentDigest: jest.fn().mockReturnValue('createContentDigest'),
  schema: {
    buildObjectType: jest.fn().mockImplementation(x => x),
  },
}

test('creates ArcGisFeature nodes', async () => {
  const pluginOptions = {
    name: 'test-source',
    url: 'https://example.com',
  }

  await plugin.sourceNodes(gatsbyContext, pluginOptions)

  expect(got).toHaveBeenCalledWith('https://example.com/0/query', {
    json: true,
    query: { f: 'geojson', outFields: '*', where: '1=1' },
  })

  expect(gatsbyContext.actions.createTypes).toMatchSnapshot()
  expect(gatsbyContext.actions.createNode).toMatchSnapshot()
})

test('supports MultiPolygon features with additional fields', async () => {
  const pluginOptions = {
    name: 'test-source',
    url: 'https://example.com',
  }

  await plugin.sourceNodes(gatsbyContext, pluginOptions)

  expect(got).toHaveBeenCalledWith('https://example.com/0/query', {
    json: true,
    query: { f: 'geojson', outFields: '*', where: '1=1' },
  })

  expect(gatsbyContext.actions.createTypes).toMatchSnapshot()
  expect(gatsbyContext.actions.createNode).toMatchSnapshot()
})

test('allows setting/overriding query params from plugin options', async () => {
  const pluginOptions = {
    name: 'test-source',
    url: 'https://example.com',
    params: { where: '1=0', foo: 'bar' },
  }

  await plugin.sourceNodes(gatsbyContext, pluginOptions)

  expect(got).toHaveBeenCalledWith('https://example.com/0/query', {
    json: true,
    query: { f: 'geojson', outFields: '*', where: '1=0', foo: 'bar' },
  })
})
