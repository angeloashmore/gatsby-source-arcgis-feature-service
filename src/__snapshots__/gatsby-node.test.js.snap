// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates ArcGisFeature nodes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
    \\"A GeoJSON feature from an ArcGIS feature service.\\"
    type ArcGisFeature implements Node @dontInfer {
      \\"The feature's ID within the ArcGIS feature service.\\"
      featureId: ID!

      \\"GeoJSON geometry data. Child fields do **not** need to be queried individually.\\"
      geometry: JSON

      \\"If the feature is a polygon, this is the optimal point within the polygon for a label.\\"
      polylabel: [Float!]

      \\"If provided in the plugin options, this is the name given to the plugin to categorize multiple feature services.\\"
      sourceName: String

      \\"The feature's GeoJSON type.\\"
      type: String!

      \\"The feature's GeoJSON type.\\"
      properties: ArcGisFeatureProperties @link
    }
  ",
    ],
    Array [
      "
    \\"ArcGIS GeoJSON feature properties.\\"
    type ArcGisFeatureProperties implements Node @infer {
      \\"If provided in the plugin options, this is the name given to the plugin to categorize multiple feature services.\\"
      sourceName: String
    }
  ",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`creates ArcGisFeature nodes 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "featureId": "id",
        "id": "createNodeId",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "ArcGisFeature",
        },
        "polylabel": Array [
          5,
          5,
        ],
        "properties": "createNodeId",
        "sourceName": "test-source",
        "type": undefined,
      },
    ],
    Array [
      Object {
        "bar": "baz",
        "foo": "bar",
        "id": "createNodeId",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "ArcGisFeatureProperties",
        },
        "sourceName": "test-source",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
